geomesa-hbase/geomesa-hbase-datastore/src/main/java/org/locationtech/geomesa/hbase/filters/KryoLazyDensityFilter.java

```diff
@@ -0,0 +1,379 @@
+        int width = getDensityWidth(hints);
+        int height = getDensityHeight(hints);
+        String weight = getDensityWeight(hints);
+        options.put(WEIGHT_OPT, weight);
+        options.put(ENVELOPE_OPT, envelope);
+        options.put(GRID_WIDTH_OPT, width);
+        options.put(GRID_HEIGHT_OPT, height);
+        gridSnap = new GridSnap(envelope, width, height);
+        int index = sft.indexOf((String) options.get(WEIGHT_OPT));
+        weightIndex = index;
+        options.put(WEIGHT_INDEX, weightIndex);
+    }
+
+    private void configureSFT() {
+        sft = SimpleFeatureTypes.createType("densitySft", sftString);
+        serializer = new KryoFeatureSerializer(sft, SerializationOptions.withoutId());
+    }
+
+    private String getGeomField() {
+        GeometryDescriptor gd = sft.getGeometryDescriptor();
+        if (gd == null) {
+            return null;
+        } else {
+            return gd.getLocalName();
+        }
+    }
+
+    private int getGeomIndex() {
+        return sft.indexOf(getGeomField());
+    }
+
+    private Envelope getDensityEnvelope(Hints hints) {
+        Envelope e = (Envelope) hints.get(QueryHints.DENSITY_BBOX());
+        return e;
+    }
+
+    private int getDensityWidth(Hints hints) {
+        int width = (int) hints.get(QueryHints.DENSITY_WIDTH());
+        return width;
+    }
+
+    private int getDensityHeight(Hints hints) {
+        int height = (int) hints.get(QueryHints.DENSITY_HEIGHT());
+        return height;
+    }
+
+    private String getDensityWeight(Hints hints) {
+        String weight = (String) hints.get(QueryHints.DENSITY_WEIGHT());
+        return weight;
+    }
+
+
+    /**
+     * Gets the weight for a feature from a double attribute
+     */
+    private double getWeightFromDouble(int i, SimpleFeature sf) {
+        Double d = (Double) sf.getAttribute(i);
+        if (d == null) {
+            return 0.0;
+        } else {
+            return d;
+        }
+    }
+
+    /**
+     * Tries to convert a non-double attribute into a double
+     */
+    private double getWeightFromNonDouble(int i, SimpleFeature sf) {
+        Object d = sf.getAttribute(i);
+        if (d == null) {
+            return 0.0;
+        } else {
+            Double converted = Converters.convert(d, Double.class);
+            if (converted == null) {
+                return 1.0;
+            } else {
+                return converted;
+            }
+        }
+    }
+
+    /**
+     * Evaluates an arbitrary expression against the simple feature to return a weight
+     */
+    private double getWeightFromExpression(Expression e, SimpleFeature sf) {
+        Double d = e.evaluate(sf, Double.class);
+        if (d == null) {
+            return 0.0;
+        } else {
+            return d;
+        }
+    }
+
+    /*
+     * Returns the weight of a simple feature
+     */
+    double weightFn(SimpleFeature sf) {
+        double returnVal = 0.0;
+        if (weightIndex == -2) {
```

Comment:
in scala these are fuctional expressions, so they don't get evaluated every time. If we're sticking with java, maybe we can use lambdas here.